File: leetcode082.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None:
            return head

        temp = ListNode(-1)
        temp.next = head

        pos = temp
        last = head
        flag = 0

        while last:
            flag = 0
            while last.next != None and last.val == last.next.val:
                flag = 1
                last.next = last.next.next

            if flag:
                pos.next = last.next
            else:
                pos=last

            last = last.next

        return temp.next

File: leetcodec++/leetcode082.cpp
//
// Created by 黄敏辉 on 2019-07-29.
//

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode prehead(-1);
        prehead.next = head;
        ListNode *tmp = &prehead;
        ListNode *del;
        while (tmp->next){
            if (tmp->next->next && tmp->next->val == tmp->next->next->val){
                while (tmp->next->next && tmp->next->val == tmp->next->next->val){
                    del = tmp->next->next;
                    tmp->next->next = tmp->next->next->next;
                    delete del;
                }
                del = tmp->next;
                tmp->next = tmp->next->next;
                delete del;
                continue;
            }
            tmp = tmp->next;
        }
        return prehead.next;
    }
};